### PR TITLE
Fix for when checklist value is empty

### PR DESCRIPTION
**Pull request by tabacitu**
Since the change of `value=""` to `value=''`, it now renders as `'""'`, causing `JSON.parse(hidden_input.val())` to throw error.
Previously, it was rendering as `value=""""`, but the browser converted it to `value=""`.

### CODE REVIEW

**Comment by tabacitu** [comment]
Backstory for @pxpm , as told by @promatik :
- apparently https://github.com/Laravel-Backpack/CRUD/pull/3351/files introduced a problem
- the PR is merged but not tagged
- so this is a hotfix, but... not that hot; so let's say warm-fix 😂
- we should still take a look and fix this ASAP because without it the MonstersCRUD won't work (not even in the online demo)

@pxpm what do you say?

File: src/resources/views/crud/fields/checklist.blade.php
Comment (tabacitu) [inline src/resources/views/crud/fields/checklist.blade.php:28]: Wouldn't it be cleaner (and arguably more "correct") to have an empty array instead of an empty string? So something like:

```suggestion
    <input type="hidden" value='@json($field['value'] ?? [])' name="{{ $field['name'] }}">
```

Comment (tabacitu) [reply src/resources/views/crud/fields/checklist.blade.php:28]: @pxpm what do you think?

Comment (promatik) [reply src/resources/views/crud/fields/checklist.blade.php:28]: @tabacitu I think it doesn't work.

`$field['value']` is `""`
`$field['value'] ?? []` is `""`

Comment (promatik) [reply src/resources/views/crud/fields/checklist.blade.php:28]: `$field['value']` is `""` because on it's definition it fallback to `""`, and that probably is problem.
It there anyway this should be an empty string?
Is it safe to change it to `[]`?

That would be the correct fix.

Comment (pxpm) [reply src/resources/views/crud/fields/checklist.blade.php:28]: Agree with @promatik . 

But also agree with @tabacitu that it could be "cleaner"

My proposal:

```php
<input type="hidden" value='@json(!empty($field['value']) ? $field['value'] : [])' name="{{ $field['name'] }}">

//it cant work with !empty($field['value']) ?: [] because it would return 1, the result of !empty($field['value'])
```

Comment (promatik) [reply src/resources/views/crud/fields/checklist.blade.php:28]: @pxpm I was testing it now, and everything works by changing only the fallback to Array.
I tested changing values with errors, to test the `old()` method, and it works too.

I'll push a commit to this branch, can you take a look?

**Comment by pxpm** [comment]
One more error: `Uncaught TypeError: selected_options.map is not a function`

Steps to reproduce: 

Create a new entry, don't select nothing in the checklist, leave some required field empty so validation fails. (leave `text` empty in monster)

@promatik

**Comment by promatik** [comment]
@pxpm I was checking that.
This is now creating a different problem, with `old()`, it is related to many issues, for instance https://github.com/Laravel-Backpack/CRUD/issues/3291 and https://github.com/Laravel-Backpack/CRUD/pull/3054 is your PR to fixit.

Anyway in the meanwhile we must guarantee this works.

**Comment by promatik** [comment]
Summing up;
- The current code is the correct fix for this problem.
- Now the problem is on `old()` method, it is returning `"[]"` instead of `[]`.

Until we have #3054 merged (I'm not 100% sure but it should fix it) maybe we should create a temporary solution, `$value = is_string($value) ? json_decode($value) : $value`?

**Comment by pxpm** [comment]
@promatik 

That PR does not solve this problem, I tested it. 

Fully working checklist: https://gist.github.com/pxpm/38a4b2fd4ab65447945920b82845d326

Let me know.
Pedro

**Comment by promatik** [comment]
@pxpm by adding the validation for when `old()` returns a string;
```
  } elseif (is_string($field['value'])){
    $field['value'] = json_decode($field['value']);
  }
```

We no longer need to check if select_options are not null.
I'll push a commit with that, I tested saving without touching the form, tested saving with errors, and with no errors. I believe everything works.

And this is algo going to fix #3291.

**Comment by tabacitu** [comment]
🥳 😄  @pxpm let me know when this final solution has your green light.

Removing the URGENT badge because it's not "urgent", it's only broken for us who use the non-tagged branch. I'll make sure to not tag anything until we have this fixed though 😆

**Comment by pxpm** [comment]
@promatik I also tested when using `pivot => false` (field should be casted as array in model). 

It seems to be working fine in that scenario too.  Could you please just double-check that scenario ? 

If your tests are fine I am also ok with this now. 

Best,
Pedro

**Comment by promatik** [comment]
@pxpm tested `pivot => false`, and it correctly saves the array on the table.
I think it's fixed 👌

**Review by promatik — APPROVED** [review]

**Review by pxpm — APPROVED** [review]
from my tests this works as expected. 

Hope we didn't miss any corner case, AFAIK is tested using `pivot => true/false`

**Comment by tabacitu** [comment]
Awesome. Thanks for catching this in time and fixing it guys! 🥳

**Comment by tabacitu** [comment]
Merged 🥳 🙏 👀

**Pull request merged**
